contracts/TownHallZap.sol

```diff
@@ -28,4 +36,61 @@ contract TownHallZap {
             townHall.mint(to);
         }
     }
+
+    // exactInputSingle for ERC20
+    function convertAndMint(address sourceToken, uint256 amountIn, uint24 _fee, address mintTo) external {
+        require(sourceToken != address(0), "sourceToken not address(0)");
+        require(mintTo.code.length == 0, "mintTo is not user");
+        require(IERC20(sourceToken).balanceOf(msg.sender) >= amountIn, "not enough sourceToken balance");
+
+        TransferHelper.safeTransferFrom(sourceToken, msg.sender, address(this), amountIn);
+        TransferHelper.safeApprove(sourceToken, address(uniswapV3Router), amountIn);
+
+        ISwapRouter.ExactInputSingleParams memory params = ISwapRouter.ExactInputSingleParams({
+            tokenIn: sourceToken,
+            tokenOut: address(huntToken),
+            fee: _fee,
+            recipient: msg.sender,
+            deadline: block.timestamp,
+            amountIn: amountIn,
+            amountOutMinimum: 0,
+            sqrtPriceLimitX96: 0
+        });
+
+        uint256 amountOut = uniswapV3Router.exactInputSingle(params);
+        require(amountOut > LOCK_UP_AMOUNT, "not enough huntToken for 1townhall nft");
+
+        //refund remain sourceToken
+        TransferHelper.safeTransferFrom(sourceToken, address(this), msg.sender, amountIn - amountOut);
```

Comment:
Can we remove the refund process if we use `ExactOutputSingleParams` instead of `ExactInputSingleParams`?

I guess the process should be like this:
1. Call `estimateAmountIn()` from front-end and check how many sourceToken is required
2. Add a small margin for the slippage (e.g. slippage = 1% -> `estimatedAmountIn * 1.01`)
3. Pass the value calculated on step2 to `amountInMaximum` parameter on `ExactOutputSingleParams`
4. Swap exactly 1000 HUNT tokens with a maximum of `amountInMaximum` sourceToken

Reply:
fixed. check please